docs/adr/2023-08-18_025-resource-based-api.md

```diff
@@ -0,0 +1,146 @@
+---
+slug: 25
+title: |
+  25. Event-sourced, resource-based API
+authors: []
+tags: [Draft]
+---
+
+## Status
+
+Draft
+
+## Context
+
+- [ADR-3](/adr/3) concluded that a full-duplex communication channels are
+  desirable to interact with a _reactive_ system.
+
+- The Client API communicates several types of messages to clients. Currently
+  this ranges from node-level `PeerConnected`, over head-specific `HeadIsOpen`
+  to messages about transactions like `TxValid`. These messages are all of type
+  `ServerOutput`.
+
+- Current capabilities of the API:
+
+  - Clients can retrieve the whole history of `ServerOutput` messages or
+    opt-out using a query parameter - all or nothing.
+
+  - There is a welcome message called `Greetings` which is always sent, that
+    contains the last `headStatus`.
+
+  - There exists a `GetUTxO` query-like `ClientInput`, which will respond with a
+    `GetUTxOResponse` containing the confirmed UTxO in an open head, or (!) the
+    currently committed UTxO when the head is initializing.
+
+  - While overall `json` encoded, clients can choose choose between `json` or
+    binary (`cbor`) output of `transaction` fields in several of these using a
+    query parameter.
+
+- Many of these features have been added in a "quick and dirty" way, by monkey
+  patching the encoded JSON.
+
+- The current capabalities even do not satisfy all user needs:
+
+  - Need to wade through lots of events to know the latest state (except the
+    very basic `headStatus` from the `Greetings`).
+
+  - Need to poll `GetUTxO` _or_ aggregate confirmed transactions on client side
+    to know the latest UTxO for constructing transactions.
+
+  - Inclusion of the whole UTxO in the head is not always desirable and
+    filtering by address would be beneficial. (not addressed in this ADR though)
+
+  - As [ADR-15](/adr/15) also proposes, some clients may not need (or should
+    not have) access to administrative information.
+
+- It is often a good idea to separate the responsibilities of Commands and
+  Queries (CQRS), as well as the model they use.
+
+## Decision
+
+- Do not add a general purpose querying interface (e.g. graphql) to the
+  `hydra-node` and stay true to [ADR-3](/adr/3) of using duplex communication
+  channels (using websockets).
+
+- **Drop** `GetUTxO` and `GetUTxOResponse` messages as they advocate a
+  request/response way of querying.
+
+- Realize that `ClientInput` data is actually a `ClientCommand` (renaming them)
+  and that `ServerOutput` are just `projections` of the [internal event stream
+  (see ADR-24)](/adr/24) into read `models` on the API layer.
+
+- Compose the API out of resource `models`, which compartmentalize the domain
+  into topics on the API layer.
+
+  - Each resource has a _latest_ state, which can be queried or clients may
+    subscribe to any model changes.
```

Comment:
I would separate that in 2 statements to make it clearer there are 2 different things, eg. :
* Clients can synchronously query the _latest_ state of each resource
* Clients can _subscribe_ to be notified of state changes for each resource

Note this definitely looks very similar to what GraphQL provides :)

Reply:
Are subscriptions in scope of standard graphql? Couldn't see it in most implementations and only apollo supports that.

It's quite inspired by graphql and maybe just doing that is simpler? Happy to discuss!